README.md
# NAME

Moonshine::Component - HTML Component base.

# VERSION

Version 0.01

# SYNOPSIS

    package Moonshine::Component::Glyphicon;

    use Moonshine::Util qw/join_class prepend_str/;

    our @ISA; { @ISA = 'Moonshine::Component' };

    BEGIN { 
        my %modifer_spec = map { $_ => 0 } qw/switch switch_base/;
        %HAS = (
            %Moonshine::Component::HAS,
            modifier_spec => sub { \%modifer_spec }
        );
    }

    sub modify {
        my $self = shift;
        my ($base_args, $build_args, $modify_args) = @_;
        if (my $class = join_class($modify_args->{switch_base}, $modify_args->{switch})){
            $base_args->{class} = prepend_str($class, $base_args->{class});
        }
        return $base_args, $build_args, $modify_args;
    }

    sub glyphicon {
        my $self = shift;
        my ( $base_args, $build_args ) = $self->validate_build(
            {
                params => $_[0] // {},
                spec => {
                    switch      => 1,
                    switch_base => { default => 'glyphicon glyphicon-' },
                    aria_hidden => { default => 'true' },
                }
            }
        );
        return $self->span($base_args);
    }

# AUTHOR

LNATION, `<thisusedtobeanemail at gmail.com>`

# BUGS

Please report any bugs or feature requests to `bug-moonshine-world at rt.cpan.org`, or through
the web interface at [http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Moonshine-World](http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Moonshine-World).  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

# ACKNOWLEDGEMENTS

# LICENSE AND COPYRIGHT

Copyright 2017 LNATION.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

[http://www.perlfoundation.org/artistic\_license\_2\_0](http://www.perlfoundation.org/artistic_license_2_0)

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changes
Revision history for Moonshine-Component

0.01    Date/time
        First version, released on an unsuspecting world.

0.02    30/01/2017
        Adds Moonshine::Util to build_requires - Hopefully that fixes all the failing tests.

0.03    31/01/2017
        Magic the synopsis.

0.04    01/02/2017
        Remove things that should not be there.

0.05    05/02/2017
        Fix Dependencies and MIN_PERL_VERSION.


t/inc/Dreams.pm
package Dreams;

use warnings;
use strict;

use Test::More;
use Params::Validate qw/:all/;
use B qw/svref_2object/;
use Scalar::Util qw/blessed/;
use base 'Test::Builder::Module';
use Exporter 'import';

use feature qw/switch/;
no if $] >= 5.017011, warnings => 'experimental::smartmatch';
use Scalar::Util qw/reftype blessed/;

our @EXPORT = qw/ moon_test render_me moon_one_test done_testing/;

our %EXPORT_TAGS =
  ( all => [qw/moon_test render_me moon_one_test done_testing/] );

my $tb = __PACKAGE__->builder;

=head1 moon_test
    
    my $args = {
        build => {
             class => 'Moonshine::Element', 
             args => {
                tag => 'p',
                text => 'hello'
             }
        }
        instructions => [
           { 
                action => 'render'
                args => {

                } 
                expected => '<p>hello</p>'
           },
           {
               action => 'text',
               expected => 'hello',
           },
        ],
    }

=cut

sub moon_test {
    my $args = shift;

    my $instructions = $args->{instructions};

    my $class;
    unless ( $class = $args->{class} ) {
        my $build = $args->{build};
        $class = $build->{class}->new( $build->{args} // {} );
    }

    for my $instruction ( @{$instructions} ) {

        my $action   = $instruction->{action};
        my $expected = $instruction->{expected};

        $action && $expected
          or diag explain $instruction;

        my $test =
          defined $instruction->{args}
          ? $class->$action( $instruction->{args} )
          : $class->$action;

        if ( my $blessed = blessed $test) {
            $tb->is_eq( $blessed, $expected,
                "$action returns Blessed - $expected" );

            if ( my $subtests = $instruction->{sub_tests} ) {
                diag 'Build new args for sub_tests';
                my $new_args = {
                    class        => $test,
                    instructions => $subtests,
                };
                moon_test($new_args);
                diag 'Return to reality';
            }
        }
        else {
            given ( reftype \$test ) {
                when (/REF/) {
                    $tb->is_deeply( $test, $expected, "$action IS DEEPLY - " );
                    diag explain $test;
                }
                when (/SCALAR/) {
                    $tb->is_eq( $test, $expected,
                        "$action IS SCALAR - $expected" );
                }
                default {
                    die diag explain $test;
                }
            }
        }
    }
}

=head2 render_me

    render_me(
        instance => , 
        action => '',
        args => { },
        expected => '<p>hello</p>',
    )


=cut

sub render_me {
    my %args = validate_with(
        params => \@_,
        spec   => {
            instance => 1,
            action   => 0,
            args     => { default => {} },
            expected => { type => SCALAR },
        }
    );

    my $instance;
    if (my $action = $args{action}) {
        $instance = $args{instance}->$action( $args{args} );   
    } else {
        $instance = $args{instance};
    }

    return $tb->is_eq( $instance->render,
        $args{expected}, "rendered - $args{expected}" );
}

sub moon_one_test {
    my %instruction = validate_with(
        params => \@_,
        spec   => {
            instance  => 0,
            meth      => 0,
            action    => 0,
            args      => { default => {} },
            args_list => 0,
            test      => 1,
            expected  => 1,
        }
    );

    my @expected = $instruction{expected};

    my @test;
    my $test_name;

    if ($instruction{test} eq 'CATCH') {
        eval { _prepare_test(%instruction) };
        @test = $@;
    }
    else {
        @test = _prepare_test(%instruction);
        $test_name = shift @test;
    }

    given ( $instruction{test} ) {
        when (/REF/) {
            return is_deeply( $test[0], $expected[0], "$test_name IS DEEP" );
        }
        when (/SCALAR/) {
            my $txt = $expected[0] // 'undef';
            return $tb->is_eq( $test[0], $expected[0],
                "$test_name IS SCALAR - $txt" );
        }
        when (/HASH/) {
            return is_deeply( {@test}, {@expected}, "$test_name IS DEEP" );
        }
        when (/ARRAY/) {
            return is_deeply( \@test, \@expected, "$test_name IS DEEP" )
        }
        when (/OBJ/) {
            return $tb->is_eq(
                blessed $test[0][0],
                $expected[0],
                "$test_name returns Blessed - $expected[0]"
            );
        }
        when (/CATCH/) {
            return $tb->like(
                $test[0],
                $expected[0],
                "catches something like - $expected[0]",
            );
        }
        when (/RENDER/) {
            return render_me(
                instance => $test[0][0],
                expected => $expected[0],
            );
        }
        default {
            die diag explain \@test;
        }
    }
}

sub _prepare_test {
    my %instruction = @_;

    my $test_name;
    if (my $action = $instruction{action}) {
        $test_name = $action;
        return defined $instruction{args_list}
          ? ($test_name, $instruction{instance}->$action(@{ $instruction{args} }))
          : ($test_name, $instruction{instance}->$action( $instruction{args} ));
    }
    elsif(my $meth = $instruction{meth}) {
        my $cv = svref_2object($meth);
        my $gv = $cv->GV;
        my $test_name = $gv->NAME;
        return defined $instruction{args_list}
          ? ($test_name, $meth->( @{ $instruction{args} } ))
          : ($test_name, $meth->( $instruction{args} ));
    }
    else {
        diag explain %instruction;
        die;
    }
}




lib/Moonshine/Component.pm
package Moonshine::Component;

use strict;
use warnings;

use Moonshine::Element;
use Moonshine::Magic;
use Params::Validate qw(:all);
use Ref::Util qw(:all);
use feature qw/switch/;
no if $] >= 5.017011, warnings => 'experimental::smartmatch';

extends "UNIVERSAL::Object";

our $VERSION = '0.06';

BEGIN {
    my $fields = $Moonshine::Element::HAS{"attribute_list"}->();
    my %html_spec = map { $_ => 0 } @{$fields}, qw/data tag/;

    has (
        html_spec     => sub { \%html_spec },
        modifier_spec => sub { { } },
    );
}

sub validate_build {
    my $self = shift;
    my %args = validate_with(
        params => $_[0],
        spec   => {
            params => { type => HASHREF },
            spec   => { type => HASHREF },
        }
    );

    my %html_spec   = %{ $self->html_spec };
    my %html_params = ();

    my %modifier_spec   = %{ $self->modifier_spec };
    my %modifier_params = ();

    my %combine = ( %{ $args{params} }, %{ $args{'spec'} } );

    for my $key ( keys %combine ) {
        my $spec = $args{spec}->{$key};
        if ( is_hashref($spec) ) {
            defined $spec->{build} and next;
            if ( exists $spec->{base} ) {
                my $param = delete $args{params}->{$key};
                my $spec  = delete $args{spec}->{$key};
                $html_params{$key} = $param if $param;
                $html_spec{$key}   = $spec  if $spec;
                next;
            }
        }

        if ( exists $html_spec{$key} ) {
            if ( defined $spec ) {
                $html_spec{$key} = delete $args{spec}->{$key};
            }
            if ( exists $args{params}->{$key} ) {
                $html_params{$key} = delete $args{params}->{$key};
            }
            next;
        }
        elsif ( exists $modifier_spec{$key} ) {
            if ( defined $spec ) {
                $modifier_spec{$key} = delete $args{spec}->{$key};
            }
            if ( exists $args{params}->{$key} ) {
                my $params = $args{params}->{$key};
                $modifier_params{$key} = delete $args{params}->{$key};
            }
            next;
        }
    }

    my %base = validate_with(
        params => \%html_params,
        spec   => \%html_spec,
    );

    my %build = validate_with(
        params => $args{params},
        spec   => $args{spec},
    );

    my %modifier = validate_with(
        params => \%modifier_params,
        spec   => \%modifier_spec,
    );

    for my $element (qw/before_element after_element children/) {
        if ( defined $modifier{$element} ) {
            my @elements = $self->build_elements( @{ $modifier{$element} } );
            $base{$element} = \@elements;
        }
    }
    
    if (scalar keys %modifier and $self->can('modify')) {
        return $self->modify(\%base, \%build, \%modifier);
    }

    return \%base, \%build, \%modifier;
}

sub build_elements {
    my $self = shift;
    my @elements_build_instructions = @_;

    my @elements;
    for my $build (@elements_build_instructions) {
        my $element;
        if ( is_blessed_ref($build) and $build->isa('Moonshine::Element') ) {
            $element = $build;
        }
        elsif ( my $action = delete $build->{action} ) {
            $self->can($action) and $element = $self->$action($build)
              or die "cannot find action - $action";
        }
        elsif ( defined $build->{tag} ) {
            $element = Moonshine::Element->new($build);
        }
        else {
            my $error_string =
              join( ", ", map { "$_: $build->{$_}" } keys %{$build} );
            die sprintf "no instructions to build the element: %s",
              $error_string;
        }
        push @elements, $element;
    }

    return @elements;
}
            
1;

__END__

=head1 NAME

Moonshine::Component - HTML Component base.

=head1 VERSION

Version 0.06

=cut

=head1 SYNOPSIS

    package Moonshine::Component::Glyphicon;

    use Moonshine::Util qw/join_class prepend_str/;

    extends 'Moonshine::Component';
    
    lazy_components (qw/span/)
    
    has (
        modifier_spec => sub { 
            {
                switch => 0,
                switch_base => 0,
            }
        }
    );

    sub modify {
        my $self = shift;
        my ($base_args, $build_args, $modify_args) = @_;
        if (my $class = join_class($modify_args->{switch_base}, $modify_args->{switch})){
            $base_args->{class} = prepend_str($class, $base_args->{class});
        }
        return $base_args, $build_args, $modify_args;
    }

    sub glyphicon {
        my $self = shift;
        my ( $base_args, $build_args ) = $self->validate_build(
            {
                params => $_[0] // {},
                spec => {
                    switch      => 1,
                    switch_base => { default => 'glyphicon glyphicon-' },
                    aria_hidden => { default => 'true' },
                }
            }
        );
        return $self->span($base_args);
    }

=head1 AUTHOR

LNATION, C<< <thisusedtobeanemail at gmail.com> >>

=head1 BUGS

=head1 ACKNOWLEDGEMENTS

=head1 LICENSE AND COPYRIGHT

Copyright 2017 LNATION.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1;


Makefile.PL
use 5.010;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use Module::Metadata;

WriteMakefile(
    NAME             => 'Moonshine::Component',
    AUTHOR           => q{Robert Acock <thisusedtobeanemail>},
    VERSION_FROM     => 'lib/Moonshine/Component.pm',
    ABSTRACT_FROM    => 'lib/Moonshine/Component.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.010,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
        'Moonshine::Test' => 0.10,
        'Moonshine::Util' => 0.04,
        'Moonshine::Element' => 0.06,
        'Moonshine::Magic' => 0.06,
    },
    PREREQ_PM => {
        'Moonshine::Element' => 0.06,
        'Moonshine::Magic'   => 0.06,
        'Params::Validate'   => 1.26,
        'Ref::Util'          => 0.113,
        'UNIVERSAL::Object'  => 0.04, 
        'strict'             => 0,
        'warnings'           => 0,
        'feature'            => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Moonshine-Component-*' },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/ThisUsedToBeAnEmail/Moonshine-Component.git',
                web  => 'https://github.com/ThisUsedToBeAnEmail/Moonshine-Component',
            },
        },
    },
    META_ADD => {
        provides => Module::Metadata->provides(version => '2', dir => 'lib'),
    }, 
);
